File: CS/Security.Extensions/RegisterUserParameters.cs
using System;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Model;
using DevExpress.Persistent.Validation;
using DevExpress.Persistent.Base.Security;
using DevExpress.ExpressApp.DC;

namespace Security.Extensions {
    [DomainComponent]
    [ModelDefault("Caption", "Register User")]
    [ImageName("BO_User")]
    public class RegisterUserParameters : ILogonActionParameters {
        public const string ValidationContext = "RegisterUserContext";
        [RuleRequiredField(null, ValidationContext)]
        public string UserName { get; set; }
        [ModelDefault("IsPassword", "True")]
        public string Password { get; set; }
        [RuleRequiredField(null, ValidationContext)]
        [RuleRegularExpression(null, ValidationContext, ManageUsersOnLogonController.EmailPattern)]
        public string Email { get; set; }
        public void Process(IObjectSpace objectSpace) {
            IAuthenticationStandardUser user = objectSpace.FindObject(SecurityExtensionsModule.SecuritySystemUserType, new BinaryOperator("UserName", UserName)) as IAuthenticationStandardUser;
            if (user != null)
                throw new ArgumentException("The login with the entered UserName or Email was already registered within the system");
            else
                SecurityExtensionsModule.CreateSecuritySystemUser(objectSpace, UserName, Email, Password, false);
            //throw new UserFriendlyException("A new user has successfully been registered");
        }
    }
}

File: VB/Security.Extensions/RegisterUserParameters.vb
Imports System
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Model
Imports DevExpress.Persistent.Validation
Imports DevExpress.Persistent.Base.Security
Imports DevExpress.ExpressApp.DC

Namespace Security.Extensions
	<DomainComponent, ModelDefault("Caption", "Register User"), ImageName("BO_User")>
	Public Class RegisterUserParameters
		Implements ILogonActionParameters

		Public Const ValidationContext As String = "RegisterUserContext"
		<RuleRequiredField(Nothing, ValidationContext)>
		Public Property UserName() As String
		<ModelDefault("IsPassword", "True")>
		Public Property Password() As String
		<RuleRequiredField(Nothing, ValidationContext), RuleRegularExpression(Nothing, ValidationContext, ManageUsersOnLogonController.EmailPattern)>
		Public Property Email() As String
		Public Sub Process(ByVal objectSpace As IObjectSpace) Implements ILogonActionParameters.Process
			Dim user As IAuthenticationStandardUser = TryCast(objectSpace.FindObject(SecurityExtensionsModule.SecuritySystemUserType, New BinaryOperator("UserName", UserName)), IAuthenticationStandardUser)
			If user IsNot Nothing Then
				Throw New ArgumentException("The login with the entered UserName or Email was already registered within the system")
			Else
				SecurityExtensionsModule.CreateSecuritySystemUser(objectSpace, UserName, Email, Password, False)
			End If
			'throw new UserFriendlyException("A new user has successfully been registered");
		End Sub
	End Class
End Namespace
